llpc/translator/lib/SPIRV/SPIRVReader.cpp

```diff
@@ -1214,352 +1257,259 @@ bool SPIRVToLLVM::postProcessRowMajorMatrix() {
-              Type *const newLoadElemType =
-                  FixedVectorType::get(castType->getArrayElementType(), castType->getArrayNumElements());
-              const unsigned addrSpace = pointerElem->getType()->getPointerAddressSpace();
-              castType = newLoadElemType->getPointerTo(addrSpace);
-              pointerElem = getBuilder()->CreateBitCast(pointerElem, castType);
-              // TODO: Remove this when LLPC will switch fully to opaque pointers.
-              assert(IS_OPAQUE_OR_POINTEE_TYPE_MATCHES(pointerElem->getType(), newLoadElemType));
-
-              LoadInst *const newLoadElem =
-                  getBuilder()->CreateAlignedLoad(newLoadElemType, pointerElem, load->getAlign(), load->isVolatile());
-              newLoadElem->setOrdering(load->getOrdering());
-              newLoadElem->setSyncScopeID(load->getSyncScopeID());
+              auto [ptr, align] = createGepIntoRowMajorMatrix(matrixType, matrix, load->getAlign(), row, col);
+              LoadInst *const newLoad = m_builder->CreateAlignedLoad(matrixElementType, ptr, align, load->isVolatile());
+              newLoad->setOrdering(load->getOrdering());
+              newLoad->setSyncScopeID(load->getSyncScopeID());
 
               if (load->getMetadata(LLVMContext::MD_nontemporal))
-                transNonTemporalMetadata(newLoadElem);
+                transNonTemporalMetadata(newLoad);
+
+              if (dim == 0) {
+                columnResult = newLoad;
+              } else {
+                if (isa<ArrayType>(columnType))
+                  columnResult = m_builder->CreateInsertValue(columnResult, newLoad, loadRow);
+                else
+                  columnResult = m_builder->CreateInsertElement(columnResult, newLoad, loadRow);
+              }
 
-              newLoad = getBuilder()->CreateInsertValue(newLoad, newLoadElem, i);
-            }
+              ++loadRow;
+            } while (dim >= 1 && loadRow < rowCount);
 
-            load->replaceAllUsesWith(getBuilder()->CreateTransposeMatrix(newLoad));
-          } else {
-            // Otherwise we are loading a single element and it's a simple load.
-            // TODO: Remove this when LLPC will switch fully to opaque pointers.
-            assert(IS_OPAQUE_OR_POINTEE_TYPE_MATCHES(pointer->getType(), pointerElemType));
-            LoadInst *const newLoad =
-                getBuilder()->CreateAlignedLoad(pointerElemType, pointer, load->getAlign(), load->isVolatile());
-            newLoad->setOrdering(load->getOrdering());
-            newLoad->setSyncScopeID(load->getSyncScopeID());
+            if (dim <= 1)
+              result = columnResult;
+            else
+              result = m_builder->CreateInsertValue(result, columnResult, loadCol);
 
-            if (load->getMetadata(LLVMContext::MD_nontemporal))
-              transNonTemporalMetadata(newLoad);
+            ++loadCol;
+          } while (dim >= 2 && loadCol < colCount);
```

Comment:
`dim` can never be `> 2`, right? It would make more sense to write `while (dim == 2 && ..`.